### PR TITLE
ddtrace-run: add gevent patch all feature

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
## Description

Add functionality to call gevent.monkey.patch_all() with ddtrace-run by
setting the environment variable DD_GEVENT_PATCH_ALL=true. This ensures
that gevent patching is done as early as possible in the application.

## Checklist
- [x] Entry added to release notes using [reno](https://docs.openstack.org/reno/latest/user/usage.html)
- [x] Tests provided; and/or
- [x] ~Description of manual testing performed and explanation is included in the code and/or PR.~
- [x] ~Library documentation is updated.~
- [x] ~[Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).~

### CODE REVIEW

File: ddtrace/bootstrap/sitecustomize.py [hunk 1]
```diff
@@ -2,15 +2,27 @@
...
-import imp
 import sys
 
-from ddtrace.utils.formats import asbool, get_env, parse_tags_str
-from ddtrace.internal.logger import get_logger
-from ddtrace import config, constants
-from ddtrace.tracer import debug_mode, DD_LOG_FORMAT
+# Perform gevent patching as early as possible in the application before
+# importing more of the library internals.
+if str(os.environ.get("DD_GEVENT_PATCH_ALL")).lower() in ("true", "1"):
```
Comment (jd) [inline ddtrace/bootstrap/sitecustomize.py:12]:
```suggestion
if os.environ.get("DD_GEVENT_PATCH_ALL", "false").lower() in ("true", "1"):
```

File: ddtrace/bootstrap/sitecustomize.py [hunk 2]
```diff
@@ -2,15 +2,27 @@
...
+    try:
+        import gevent.monkey
+    except ImportError:
+        print("failed to import gevent.monkey", file=sys.stderr)
+    else:
+        gevent.monkey.patch_all()
+        pass
```
Comment (jd) [inline ddtrace/bootstrap/sitecustomize.py:19]:
```suggestion
```

File: ddtrace/bootstrap/sitecustomize.py [hunk 3]
```diff
@@ -2,15 +2,27 @@
...
-from ddtrace.internal.logger import get_logger
-from ddtrace import config, constants
-from ddtrace.tracer import debug_mode, DD_LOG_FORMAT
+# Perform gevent patching as early as possible in the application before
+# importing more of the library internals.
+if str(os.environ.get("DD_GEVENT_PATCH_ALL")).lower() in ("true", "1"):
+    try:
+        import gevent.monkey
+    except ImportError:
+        print("failed to import gevent.monkey", file=sys.stderr)
```
Comment (jd) [inline ddtrace/bootstrap/sitecustomize.py:16]: does it make sense to continue at this point? you might run a broken application
I'd just leave the exception being raised.

**Review by jd — CHANGES_REQUESTED** [review]

File: ddtrace/bootstrap/sitecustomize.py (hunk 3 above)
Comment (Kyle-Verhoog) [reply ddtrace/bootstrap/sitecustomize.py:16]: Good point. Especially since the flag is explicitly passed 👍

**Review by jd — APPROVED** [review]

**Comment by majorgreys** [comment]
Should this be the default behavior when gevent is installed as well?

**Comment by Kyle-Verhoog** [comment]
@majorgreys I did give it some thought. It'd probably be fine for ~99% of use-cases but I'm concerned about the 1% that have gevent installed and it isn't being used by the application. I don't think we want to be gevent-monkey-patching people's non-gevent applications.

**Comment by jd** [comment]
Yeah that should not be the default, it'd break many applications. It's common to have gevent installed by a third-party library because it supports it and has listed it in its deps (maybe wrongly). That'd break everything.

**Review by jd — DISMISSED** [review]

**Review by majorgreys — APPROVED** [review]

**Pull request merged**